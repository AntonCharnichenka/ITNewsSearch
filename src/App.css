/* .dropdown {
    position: relative;
}

.menu {
    position: absolute;
    list-style-type: none;
    margin: 5px 0;
    padding: 0;

    border: 1px solid grey;
    width: 150px;
}

.menu > li {
    margin: 0;
    background-color: white;
}

.menu > li:hover {
    background-color: lightgray;
}

.menu > li > button {
    width: 100%;
    height: 100%;
    text-align: left;

    background: none;
    color: inherit;
    border: none;
    padding: 5px;
    margin: 0;
    font: inherit;
    cursor: pointer;
}

 */

.container {
    height: 100vw;
    padding: 20px;

    background: #83a4d4;
    background: linear-gradient(to left, #b6fbff, #83a4d4);

    color: #171212;
}

.headline-primary {
    font-size: 48px;
    font-weight: 300;
    letter-spacing: 2px;
}

.item {
    display: flex;
    align-items: center;
    padding-bottom: 5px;
}

.item > span {
    padding: 0px 5px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}

.item > span > a {
    color: inherit;
}

.button {
    background: transparent;
    border: 1px solid #171212;
    padding: 5px;
    cursor: pointer;
}

.button:hover {
    background: #171212;
    color: #ffffff;
}

.button_small {
    padding: 2px;
}

.button_large {
    padding: 10px;
}

.search-form {
    padding: 10px 0px 20px 0;
    display: flex;
    align-items: baseline;
}

.label {
    border-top: 1px solid #171212;
    border-left: 1px solid #171212;
    padding-left: 5px;
    font-size: 24px;
}

.input {
    border: none;
    border-bottom: 1px solid #171212;
    background-color: transparent;
    font-size: 24px;
}