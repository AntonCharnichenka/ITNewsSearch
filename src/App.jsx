import * as React from 'react';

function App() {
  const [text, setText] = React.useState('Some text ...');

  function handleOnChange(event) {
    setText(event.target.value);
  }

  const ref = (node) => {
    console.log('REF CALLBACK CALLED')
    
    if (!node) {
      console.log('RETURNING NOTHING FROM REF CALLBACK')
      return
    };

    console.log('REF CALLBACK CALLED WITH NODE', node)
    const { width } = node.getBoundingClientRect();

    if (width >= 150) {
      node.style.color = 'red';
    } else {
      node.style.color = 'blue';
    }
  };

  return (
    <div>
      <input type="text" value={text} onChange={handleOnChange} />
      <div>
        <span ref={ref}>{text}</span>
      </div>
    </div>
  );
}

export default App;

// create a class with a constructor with parameter 'number' and method to print it
class Number {
  constructor(number) {
    this.number = number;
  }

  print() {
    console.log(this.number);
  }
}

// create instance of this class
const number = new Number(5);