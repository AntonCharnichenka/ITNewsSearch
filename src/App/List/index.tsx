import * as React from 'react';
import { sortBy } from 'lodash';
import { CiCircleRemove } from "react-icons/ci";
// import { FaSort } from "react-icons/fa";
import { GoSortAsc, GoSortDesc } from "react-icons/go";
import { ListProps, ItemProps, Stories, SortKey, sortState, SortFunction } from './types';
import { StyledItem, StyledColumn } from './style';
import { StyledButtonSmall } from '../shared/styles';

export const List: React.FC<ListProps> = ({ list, onRemoveItem }): JSX.Element => { // or return type could also be React.ReactNode
  const [sort, setSort] = React.useState<sortState>(
    {
      sortKey: 'NONE',
      isReversed: false
    }
  );
  const handleSort = (sortKey: SortKey) => {
    const isReversed = sort.sortKey === sortKey && !sort.isReversed;
    setSort({ sortKey, isReversed });  // shorthand object initializer notation 
  };

  const Sorts: Record<SortKey, SortFunction> = {
    'NONE': (list: Stories): Stories => list,
    'TITLE': (list: Stories): Stories => sortBy(list, 'title'),
    'AUTHOR': (list: Stories): Stories => sortBy(list, 'author'),
    'COMMENTS': (list: Stories): Stories => sortBy(list, 'num_comments').reverse(),
    'POINTS': (list: Stories): Stories => sortBy(list, 'points').reverse(),
  };

  const sortFunction = Sorts[sort.sortKey];
  const sortedList = sort.isReversed ? sortFunction(list).reverse() : sortFunction(list);

  const getSortPictogram = (sortKey: SortKey) => {
    const GoSort = sort.isReversed ? GoSortDesc : GoSortAsc;
    return <GoSort color={sortKey === sort.sortKey ? 'red' : undefined} />
  };

  return <ul>
    <StyledItem>
      <StyledColumn width="40%">
        <StyledButtonSmall
          type="button"
          onClick={() => handleSort('TITLE')}>Title{getSortPictogram('TITLE')}
        </StyledButtonSmall>
      </StyledColumn>
      <StyledColumn width="25%">
        <StyledButtonSmall
          type="button"
          onClick={() => handleSort('AUTHOR')}>Author{getSortPictogram('AUTHOR')}
        </StyledButtonSmall>
      </StyledColumn>
      <StyledColumn width="15%">
        <StyledButtonSmall
          type="button"
          onClick={() => handleSort('COMMENTS')}>Comments{getSortPictogram('COMMENTS')}
        </StyledButtonSmall>
      </StyledColumn>
      <StyledColumn width="10%">
        <StyledButtonSmall
          type="button"
          onClick={() => handleSort('POINTS')}>Points{getSortPictogram('POINTS')}
        </StyledButtonSmall>
      </StyledColumn>
      <StyledColumn width="10%"></StyledColumn>
    </StyledItem>

    {sortedList.map((item) => (
      <Item
        key={item.objectID}
        item={item}
        onRemoveItem={onRemoveItem}
      />
    ))}
  </ul>
};

export const Item = ({ item, onRemoveItem }: ItemProps): JSX.Element => ( // or const Item: React.FC<ItemProps> = ({ item, onRemoveItem }) => ( ...
  <StyledItem>
    <StyledColumn width="40%">
      <a href={item.url}>{item.title}</a>
    </StyledColumn>
    <StyledColumn width="25%">{item.author}</StyledColumn>
    <StyledColumn width="15%">{item.num_comments}</StyledColumn>
    <StyledColumn width="10%">{item.points}</StyledColumn>
    <StyledColumn width="10%">
      <StyledButtonSmall type="button" onClick={() => onRemoveItem(item)}>
        <CiCircleRemove />
      </StyledButtonSmall>
    </StyledColumn>
  </StyledItem>
);
