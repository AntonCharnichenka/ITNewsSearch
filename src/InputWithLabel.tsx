import * as React from 'react';
import  { StyledInput, StyledLabel } from './styles';

type InputWithLabelProps = {
    id: string;
    value: string;
    type?: string;
    onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
    isFocused?: boolean;
    children: React.ReactNode;
  };

export const InputWithLabel: React.FC<InputWithLabelProps> = ({
    id,
    value,
    type = 'text',
    onInputChange,
    isFocused,
    children,
  }) => {
    const inputRef = React.useRef<HTMLInputElement>(null);
  
    React.useEffect(() => {
      if (isFocused && inputRef.current) {
        inputRef.current.focus();
      }
    }, [isFocused]);
  
    return (
      <>
        <StyledLabel htmlFor={id}>{children}</StyledLabel>
        &nbsp;
        <StyledInput
          ref={inputRef}
          id={id}
          type={type}
          value={value}
          onChange={onInputChange}
        />
      </>
    );
  };
  