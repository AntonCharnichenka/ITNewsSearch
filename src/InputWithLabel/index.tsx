import * as React from 'react';
import  { InputWithLabelProps } from './types';
import  { StyledInput, StyledLabel } from './style';


export const InputWithLabel: React.FC<InputWithLabelProps> = ({
    id,
    value,
    type = 'text',
    onInputChange,
    isFocused,
    children,
  }) => {
    const inputRef = React.useRef<HTMLInputElement>(null);
  
    React.useEffect(() => {
      if (isFocused && inputRef.current) {
        inputRef.current.focus();
      }
    }, [isFocused]);
  
    return (
      <>
        <StyledLabel htmlFor={id}>{children}</StyledLabel>
        &nbsp;
        <StyledInput
          ref={inputRef}
          id={id}
          type={type}
          value={value}
          onChange={onInputChange}
        />
      </>
    );
  };
  