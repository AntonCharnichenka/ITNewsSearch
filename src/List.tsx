import * as React from 'react';
import { CiCircleRemove } from "react-icons/ci";
import { StyledItem, StyledColumn, StyledButtonSmall } from './styles';

export type Story = {
  objectID: string;
  url: string;
  title: string;
  author: string;
  num_comments: number;
  points: number;
};

export type Stories = Story[];

type ItemProps = {
  item: Story;
  onRemoveItem: (item: Story) => void;
};

type ListProps = {
  list: Stories;
  onRemoveItem: (item: Story) => void;
};


export const List: React.FC<ListProps> = ({ list, onRemoveItem }): JSX.Element => ( // or return type could also be React.ReactNode
  <ul>
    {list.map((item) => (
      <Item
        key={item.objectID}
        item={item}
        onRemoveItem={onRemoveItem}
      />
    ))}
  </ul>
);

export const Item = ({ item, onRemoveItem }: ItemProps ): JSX.Element => ( // or const Item: React.FC<ItemProps> = ({ item, onRemoveItem }) => ( ...
  <StyledItem>
    <StyledColumn width="40%">
      <a href={item.url}>{item.title}</a>
    </StyledColumn>
    <StyledColumn width="40%">{item.author}</StyledColumn>
    <StyledColumn width="40%">{item.num_comments}</StyledColumn>
    <StyledColumn width="40%">{item.points}</StyledColumn>
    <StyledColumn width="40%">
      <StyledButtonSmall type="button" onClick={() => onRemoveItem(item)}>
        <CiCircleRemove/>
      </StyledButtonSmall>
    </StyledColumn>
  </StyledItem>
);
