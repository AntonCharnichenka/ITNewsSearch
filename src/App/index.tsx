import * as React from 'react';
import axios from 'axios';
// import styles from './App.module.css';
import { List } from './List';
import { Story } from './List/types';
import { SearchForm } from './SearchForm';
import { StyledContainer, StyledHeadlinePrimary } from './style';
import { StoriesAction, StoriesState, LastSearchesProps, HandleLastSearchesFunction, SearchTerm } from './types';
// import { AiOutlineLoading3Quarters } from "react-icons/ai";


export const storiesReducer = (state: StoriesState, action: StoriesAction): StoriesState => {  // return tyoe is inferred, so it might be not type annotated
  switch (action.type) {
    case 'STORIES_FETCH_INIT':
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case 'STORIES_FETCH_SUCCESS':
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload.page === 0 ? action.payload.list : state.data.concat(action.payload.list),
        page: action.payload.page,
      };
    case 'STORIES_FETCH_FAILURE':
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case 'REMOVE_STORY':
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        ),
      };
    default:
      throw new Error();
  }
};

const useStorageState = (
  key: string,
  initialState: string
): [string, (newValue: string) => void] => {
  const isMounted = React.useRef(false);

  const [value, setValue] = React.useState(  // value is inferred to be a string, setValue only takes a string as an argument
    localStorage.getItem(key) || initialState
  );

  React.useEffect(() => { // no need to add types here as they are inferred
    if (!isMounted.current) {
      isMounted.current = true;
    } else {
      localStorage.setItem(key, value);
    }
  }, [value, key]);

  return [value, setValue];
};

const getSumComments = (stories: StoriesState): number => {
  const sumComments = stories.data.reduce((result, value) => result + value.num_comments, 0);

  return sumComments
};


// const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query=';
const API_BASE = 'https://hn.algolia.com/api/v1';
const API_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page='

const getUrl = (searchTerm: string, page: number): string => `${API_BASE}${API_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}`;

const extractSearchTerm = (url: string): string => url.substring(url.lastIndexOf('?') + 1, url.lastIndexOf('&')).replace(PARAM_SEARCH, '')

const get5lastSearches = (urls: string[]): string[] => urls.reduce(
  (result: string[], url: string, index: number): string[] => {
    const searchTerm = extractSearchTerm(url);

    if (index === 0) {
      return result.concat(searchTerm);
    };

    const previousSearchTerms = result[result.length - 1];

    if (searchTerm === previousSearchTerms) {
      return result;
    } else {
      return result.concat(searchTerm);
    };
  },
  []
).slice(-6).slice(0, -1);


const App = () => {
  const [searchTerm, setSearchTerm] = useStorageState(
    'search',
    'React'
  );

  const [urls, setUrls] = React.useState(
    [getUrl(searchTerm, 0)]
  );

  const [stories, dispatchStories] = React.useReducer(
    storiesReducer,
    { data: [], page: 0, isLoading: false, isError: false }
  );

  const handleFetchStories = React.useCallback(
    async () => {
      dispatchStories({ type: 'STORIES_FETCH_INIT' });

      try {
        const lastUrl = urls[urls.length - 1];
        const result = await axios.get(lastUrl);

        dispatchStories({
          type: 'STORIES_FETCH_SUCCESS',
          payload: {
            list: result.data.hits,
            page: result.data.page,
          },
        });

      } catch {
        dispatchStories({ type: 'STORIES_FETCH_FAILURE' });
      }
    },
    [urls]
  );

  React.useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  const handleRemoveStory = (item: Story): void => {
    dispatchStories({
      type: 'REMOVE_STORY',
      payload: item,
    });
  };

  const handleSearchInput = (event: React.ChangeEvent<HTMLInputElement>) => setSearchTerm(event.target.value);

  const hadleSearch = (searchTerm: SearchTerm, page: number): void => {
    const url = getUrl(searchTerm, page);
    setUrls(urls.concat(url));
  };

  const handleSearchSubmit = (event: React.FormEvent<HTMLFormElement>): void => {
    hadleSearch(searchTerm, 0);

    event.preventDefault();
  };

  const handleLastSearch: HandleLastSearchesFunction = (searchTerm) => {
    setSearchTerm(searchTerm);
    hadleSearch(searchTerm, 0);
  };

  const handleMore = (): void => {
    const lastUrl = urls[urls.length - 1];
    const searchTerm = extractSearchTerm(lastUrl);
    hadleSearch(searchTerm, stories.page + 1);
  };

  const lastSearches = get5lastSearches(urls);

  const sumComments = React.useMemo(() => getSumComments(stories), [stories]);

  return (
    <StyledContainer>
      <StyledHeadlinePrimary>My Hacker Stories with {sumComments} comments</StyledHeadlinePrimary>
      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />
      {lastSearches.length > 0 && <LastSearches lastSearches={lastSearches} handleLastSearch={handleLastSearch} />}
      <hr />
      {stories.isError && <p>Something went wrong ...</p>}
      <List list={stories.data} onRemoveItem={handleRemoveStory} />
      {/* style button below */}
      {stories.isLoading ? (
        <p>Loading ...</p>
      ) : <button type='button' onClick={handleMore}>More</button>}
    </StyledContainer>
  );
};

const LastSearches: React.FC<LastSearchesProps> = ({ lastSearches, handleLastSearch }) => (
  <div style={{ display: "flex" }}>
    <label htmlFor="last-searches">Last searches:</label>
    &nbsp;
    <div id="last-searches">
      {
        lastSearches.map(
          (searhTerm, index) => <button key={searhTerm + index} type="button" onClick={() => handleLastSearch(searhTerm)}>{searhTerm}</button>
        )
      }
    </div>
  </div>
);

export default App;
