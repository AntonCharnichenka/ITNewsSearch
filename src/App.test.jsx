import  { describe, it, expect, vi } from 'vitest';
import { storiesReducer } from './App'
import App from './App';
import { List, Item } from './List';
import { SearchForm } from './SearchForm';
import { InputWithLabel } from './InputWithLabel';
import { render, screen, fireEvent, waitFor, within } from '@testing-library/react';
import axios from 'axios';

vi.mock('axios');

// describe(  // it is a test suite
//     'Something truthy and falsy',
//     () => {
//         it('true should be true', () => {expect(true).toBeTruthy();});  // it is a test case
//         it('false should be false', () => {expect(false).toBeFalsy();})
//     }
// );

const storyOne = {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
};

const storyTwo = {
    title: 'Redux',
    url: 'https://redux.js.org/',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
};

const stories = [storyOne, storyTwo];

describe(
    'soriesReducer',
    () => {
        it(
            'removes story from all stories if action is REMOVE_STORY',
            () => {
                const action = { type: 'REMOVE_STORY', payload: storyOne };
                const state = { data: stories, isLoading: false, isError: false };

                const new_state = storiesReducer(state, action);
                const expected_state = { data: [storyTwo], isLoading: false, isError: false};

                expect(new_state).toStrictEqual(expected_state);
            }
        );
        it(
            'set isError to true if action is STORIES_FETCH_FAILURE',
            () => {
                const action = { type: 'STORIES_FETCH_FAILURE' };
                const state = { data: [], isLoading: false, isError: false };

                const new_state = storiesReducer(state, action);
                const expected_state = { data: [], isLoading: false, isError: true };

                expect(new_state).toStrictEqual(expected_state);
            }
        );
        it(
            'set data if action is STORIES_FETCH_SUCCESS',
            () => {
                const action = { type: 'STORIES_FETCH_SUCCESS', payload: stories };
                const state = { data: [], isLoading: false, isError: false };

                const new_state = storiesReducer(state, action);
                const expected_state = { data: stories, isLoading: false, isError: false };

                expect(new_state).toStrictEqual(expected_state);
            }
        );
        it(
            'set isLoading to true if action is STORIES_FETCH_INIT',
            () => {
                const action = { type: 'STORIES_FETCH_INIT' };
                const state = { data: [], isLoading: false, isError: false };

                const new_state = storiesReducer(state, action);
                const expected_state = { data: [], isLoading: true, isError: false };

                expect(new_state).toStrictEqual(expected_state);
            }
        );
    },
);

describe(
    'Item',
    () => {
        it(
            'renders all properties',
            () => {
                render(<Item item={storyOne}/>); // renders the element in jsdom environment
                // screen.debug(); // prints the html of the rendered element
                
                expect(screen.getByText('React')).toHaveAttribute('href', 'https://reactjs.org/'); // expect needs extesion by toHaveAttribute assertion method (in setup.js) 
                expect(screen.getByText('Jordan Walke')).toBeInTheDocument(); // same as for toHaveAttribute

            }
        );
        it(
            'renders a clickable dismiss button',
            () => {
                render(<Item item={storyOne}/>);
                expect(screen.getByRole('button')).toBeInTheDocument();
            }
        );
        it(
            'clicking dismiss button calls callback handler',
            () => {
                const handleRemoveItem = vi.fn();
                render(<Item item={storyOne} onRemoveItem={handleRemoveItem}/>)
                
                fireEvent.click(screen.getByRole('button'));

                expect(handleRemoveItem).toHaveBeenCalledTimes(1);
                expect(handleRemoveItem).toHaveBeenCalledWith(storyOne);
            }
        )
        it(
            'renders snapshot',
            () => {
                const { container } = render(<Item item={storyOne}/>);
                expect(container.firstChild).toMatchSnapshot();
            }
        )

    }
);

describe(
    'SearchForm',
    () => {
        const searchFormProps = {
            searchTerm: 'React',
            onSearchInput: vi.fn(),
            onSearchSubmit: vi.fn(),
        };

        it(
            'renders input field with its value',
            () => {
                render(<SearchForm {...searchFormProps}/>);
                // screen.debug();
                expect(screen.getByDisplayValue('React')).toBeInTheDocument();
            }
        );

        it(
            'renders correct label',
            () => {
                render(<SearchForm {...searchFormProps}/>);
                expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
            }
        );

        it(
            'calls onSearchInput on input field change',
            () => {
                render(<SearchForm {...searchFormProps}/>);
                
                fireEvent.change(
                    screen.getByDisplayValue('React'),
                    { target: { value: 'Redux' }}
                );

                expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
                // expect(searchFormProps.onSearchInput).toHaveBeenCalledWith(
                //     expect.objectContaining(
                //         { target: expect.objectContaining({ value: 'Redux' }) }
                //     )
                // );
            }
        );

        it(
            'calls onSearchSubmit on button submit click',
            () => {
                render(<SearchForm {...searchFormProps}/>);
                
                fireEvent.submit(screen.getByRole('button'));

                expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
            }
        )

        it(
            'renders snapshot',
            () => {
                const { container } = render(<SearchForm {...searchFormProps}/>);
                expect(container.firstChild).toMatchSnapshot();
            }
        )
    }
);

describe(
    'List',
    () => {
        it(
            'renders items', 
            () => {
                const onRemoveItem = vi.fn();
                render(<List list={stories} onRemoveItem={onRemoveItem}/>);

                expect(screen.getAllByRole('listitem')).toHaveLength(2);

                const react_item = screen.getByText('React').closest('li');
                const redux_item = screen.getByText('Redux').closest('li');

                expect(react_item).toBeInTheDocument();
                expect(redux_item).toBeInTheDocument();

                expect(within(react_item).getByText('Jordan Walke')).toBeInTheDocument();
                expect(within(redux_item).getByText('Dan Abramov, Andrew Clark')).toBeInTheDocument();

                const react_item_button = within(react_item).getByRole('button');

                expect(react_item_button).toBeInTheDocument();
                expect(within(redux_item).getByRole('button')).toBeInTheDocument();

                fireEvent.click(react_item_button);

                expect(onRemoveItem).toHaveBeenCalledTimes(1);
                expect(onRemoveItem).toHaveBeenCalledWith(storyOne);
            }
        )
        it(
            'renders snapshot',
            () => {
                const { container } = render(<List list={stories}/>);
                expect(container.firstChild).toMatchSnapshot();
            }
        )
    }
);

describe(
    'InputWithLabel',
    () => {
        it(
            'renders input field and label for it',
            () => {
                const onInputChange = vi.fn();
                // const a = (a_event) => {
                //     const aa = 1;
                // }
                render(<InputWithLabel id='search' value='React' onInputChange={onInputChange} isFocused>Search:</InputWithLabel>)
                // screen.debug();

                expect(screen.getByLabelText(/search/i)).toBeInTheDocument();

                const input = screen.getByDisplayValue('React');
                expect(input).toBeInTheDocument();

                expect(input).toHaveFocus();
                expect(input).toHaveAttribute('id', 'search');

                fireEvent.change(input, {target: {value: 'Redux'}});
                expect(onInputChange).toHaveBeenCalledTimes(1);
                

                // expect(onInputChange).toHaveBeenCalledWith(
                //     expect.objectContaining({ target: expect.objectContaining({ value: 'Redux' }) })
                // );    
            }
        )
        // it(
        //     'renders snapshot',
        //     () => {
        //         const { container } = render(<InputWithLabel id='search' value='React' isFocused>Search:</InputWithLabel>);
        //         expect(container.firstChild).toMatchSnapshot();
        //     }
        // )

        // in the test above: wait input to be focused then make and check snapshot
    }
);

describe(
    'App',
    () => {
        it(
            'succeeds fetching data',
            async () => {
                const promise = Promise.resolve({ data: { hits: stories } });
                axios.get.mockImplementationOnce(() => promise);

                render(<App/>);

                expect(screen.getByText(/loading/i)).toBeInTheDocument();

                await waitFor(async () => await promise);

                expect(screen.queryByText(/loading/i)).toBeNull();

                expect(screen.getByText('React')).toBeInTheDocument();
                expect(screen.getByText('Redux')).toBeInTheDocument();
                expect(screen.getAllByRole('button').length).toBe(3);
            }
        );
        it(
            'fails fetching data',
            async () => {
                const promise = Promise.reject();
                axios.get.mockImplementationOnce(() => promise);

                render(<App/>);

                expect(screen.getByText(/loading/i)).toBeInTheDocument();

                try {
                    await waitFor(async () => await promise);
                } catch (error) {
                    expect(screen.queryByText(/loading/i)).toBeNull();
                    expect(screen.getByText(/something went wrong/i)).toBeInTheDocument();
                }


            }
        );
        it(
            'removes a story',
            async () => {
                const promise = Promise.resolve( {data: { hits: stories }} );
                axios.get.mockImplementationOnce(() => promise);

                render(<App/>);

                await waitFor(async () => await promise);

                expect(screen.getAllByRole('button').length).toBe(3);
                expect(screen.getByText('Jordan Walke')).toBeInTheDocument();

                fireEvent.click(screen.getAllByRole('button')[1]); // 1st button is submit for search input, 2nd is for removing 1st story

                expect(screen.getAllByRole('button').length).toBe(2);
                expect(screen.queryByText('Jordan Walke')).toBeNull();
            }
        );
        it(
            'searches for specific stories',
            async () => {
                const reactPromise = Promise.resolve({ data: { hits: stories } });
                const anotherStory = {
                    title: 'JavScript',
                    url: 'https://js.org/',
                    author: 'Brendan Eich',
                    num_comments: 15,
                    points: 10,
                    objectID: 2,
                };
                const javascriptPromise = Promise.resolve({ data: { hits: [anotherStory] } })

                axios.get.mockImplementation(
                    (url) => {
                        if (url.includes('React')) {
                            return reactPromise;
                        }
                        if (url.includes('JavScript')) {
                            return javascriptPromise;
                        }

                        throw Error();
                    }
                );

                render(<App/>);

                await waitFor(async () => await reactPromise);

                expect(screen.getByDisplayValue('React')).toBeInTheDocument();
                expect(screen.queryByDisplayValue('JavScript')).toBeNull();
                expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
                expect(screen.queryByText('Dan Abramov')).toBeNull();
                expect(screen.queryByText('Brendan Eich')).toBeNull();

                // User interaction -> Search
                fireEvent.change(
                    screen.getByDisplayValue('React'),
                    { target: { value: 'JavScript' } }
                );

                expect(screen.queryByDisplayValue('React')).toBeNull();
                expect(screen.getByDisplayValue('JavScript')).toBeInTheDocument();


                const form = screen.getByLabelText(/search/i).parentNode;
                const submit_button = form.querySelector('button');

                // Second data fetching

                fireEvent.submit(submit_button);

                await waitFor(async () => await javascriptPromise);

                expect(screen.queryByText('Jordan Walke')).toBeNull();
                expect(screen.queryByText('Dan Abramov')).toBeNull();
                expect(screen.getByText('Brendan Eich')).toBeInTheDocument();
            }
        )
    }
);
