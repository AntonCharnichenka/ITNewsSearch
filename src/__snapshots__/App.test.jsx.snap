// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Item > renders snapshot 1`] = `
<li
  class="sc-dmyCSP gbftpc"
>
  <span
    class="sc-hLQSwg ktFmUj"
    width="40%"
  >
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span
    class="sc-hLQSwg ktFmUj"
    width="40%"
  >
    Jordan Walke
  </span>
  <span
    class="sc-hLQSwg ktFmUj"
    width="40%"
  >
    3
  </span>
  <span
    class="sc-hLQSwg ktFmUj"
    width="40%"
  >
    4
  </span>
  <span
    class="sc-hLQSwg ktFmUj"
    width="40%"
  >
    <button
      class="sc-eDLKkx sc-jTQCzO jknmCz cHAImJ"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          id="Circle_Remove"
        >
          <g>
            <path
              d="M9.525,13.765a.5.5,0,0,0,.71.71c.59-.59,1.175-1.18,1.765-1.76l1.765,1.76a.5.5,0,0,0,.71-.71c-.59-.58-1.18-1.175-1.76-1.765.41-.42.82-.825,1.23-1.235.18-.18.35-.36.53-.53a.5.5,0,0,0-.71-.71L12,11.293,10.235,9.525a.5.5,0,0,0-.71.71L11.293,12Z"
            />
            <path
              d="M12,21.933A9.933,9.933,0,1,1,21.934,12,9.945,9.945,0,0,1,12,21.933ZM12,3.067A8.933,8.933,0,1,0,20.934,12,8.944,8.944,0,0,0,12,3.067Z"
            />
          </g>
        </g>
      </svg>
    </button>
  </span>
</li>
`;

exports[`List > renders snapshot 1`] = `
<ul>
  <li
    class="sc-dmyCSP gbftpc"
  >
    <span
      class="sc-hLQSwg ktFmUj"
      width="40%"
    >
      <a
        href="https://reactjs.org/"
      >
        React
      </a>
    </span>
    <span
      class="sc-hLQSwg ktFmUj"
      width="40%"
    >
      Jordan Walke
    </span>
    <span
      class="sc-hLQSwg ktFmUj"
      width="40%"
    >
      3
    </span>
    <span
      class="sc-hLQSwg ktFmUj"
      width="40%"
    >
      4
    </span>
    <span
      class="sc-hLQSwg ktFmUj"
      width="40%"
    >
      <button
        class="sc-eDLKkx sc-jTQCzO jknmCz cHAImJ"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            id="Circle_Remove"
          >
            <g>
              <path
                d="M9.525,13.765a.5.5,0,0,0,.71.71c.59-.59,1.175-1.18,1.765-1.76l1.765,1.76a.5.5,0,0,0,.71-.71c-.59-.58-1.18-1.175-1.76-1.765.41-.42.82-.825,1.23-1.235.18-.18.35-.36.53-.53a.5.5,0,0,0-.71-.71L12,11.293,10.235,9.525a.5.5,0,0,0-.71.71L11.293,12Z"
              />
              <path
                d="M12,21.933A9.933,9.933,0,1,1,21.934,12,9.945,9.945,0,0,1,12,21.933ZM12,3.067A8.933,8.933,0,1,0,20.934,12,8.944,8.944,0,0,0,12,3.067Z"
              />
            </g>
          </g>
        </svg>
      </button>
    </span>
  </li>
  <li
    class="sc-dmyCSP gbftpc"
  >
    <span
      class="sc-hLQSwg ktFmUj"
      width="40%"
    >
      <a
        href="https://redux.js.org/"
      >
        Redux
      </a>
    </span>
    <span
      class="sc-hLQSwg ktFmUj"
      width="40%"
    >
      Dan Abramov, Andrew Clark
    </span>
    <span
      class="sc-hLQSwg ktFmUj"
      width="40%"
    >
      2
    </span>
    <span
      class="sc-hLQSwg ktFmUj"
      width="40%"
    >
      5
    </span>
    <span
      class="sc-hLQSwg ktFmUj"
      width="40%"
    >
      <button
        class="sc-eDLKkx sc-jTQCzO jknmCz cHAImJ"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            id="Circle_Remove"
          >
            <g>
              <path
                d="M9.525,13.765a.5.5,0,0,0,.71.71c.59-.59,1.175-1.18,1.765-1.76l1.765,1.76a.5.5,0,0,0,.71-.71c-.59-.58-1.18-1.175-1.76-1.765.41-.42.82-.825,1.23-1.235.18-.18.35-.36.53-.53a.5.5,0,0,0-.71-.71L12,11.293,10.235,9.525a.5.5,0,0,0-.71.71L11.293,12Z"
              />
              <path
                d="M12,21.933A9.933,9.933,0,1,1,21.934,12,9.945,9.945,0,0,1,12,21.933ZM12,3.067A8.933,8.933,0,1,0,20.934,12,8.944,8.944,0,0,0,12,3.067Z"
              />
            </g>
          </g>
        </svg>
      </button>
    </span>
  </li>
</ul>
`;

exports[`SearchForm > renders snapshot 1`] = `
<form
  class="sc-iBdnpw gWhaa"
>
  <label
    class="sc-fsYfdN cmJZKY"
    for="search"
  >
    <strong>
      Search:
    </strong>
  </label>
  Â 
  <input
    class="sc-qZrbh dSzoWA"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="sc-eDLKkx sc-jTQCzO jknmCz cHAImJ"
    type="submit"
  >
    <svg
      fill="currentColor"
      height="1em"
      id="search"
      stroke="currentColor"
      stroke-width="0"
      version="1.1"
      viewBox="0 0 24 24"
      width="1em"
      x="0px"
      xmlns="http://www.w3.org/2000/svg"
      y="0px"
    >
      <g>
        <path
          d="M20.031,20.79c0.46,0.46,1.17-0.25,0.71-0.7l-3.75-3.76c1.27-1.41,2.04-3.27,2.04-5.31
		c0-4.39-3.57-7.96-7.96-7.96s-7.96,3.57-7.96,7.96c0,4.39,3.57,7.96,7.96,7.96c1.98,0,3.81-0.73,5.21-1.94L20.031,20.79z
		 M4.11,11.02c0-3.84,3.13-6.96,6.96-6.96c3.84,0,6.96,3.12,6.96,6.96c0,3.84-3.12,6.96-6.96,6.96C7.24,17.98,4.11,14.86,4.11,11.02
		z"
        />
      </g>
    </svg>
  </button>
</form>
`;
